packages/docs-app/src/examples/core-examples/treeExample.tsx

```diff
@@ -14,82 +14,116 @@
  * limitations under the License.
  */
 
-import React from "react";
+import React, { useCallback, useReducer } from "react";
 
 import { Classes, Icon, Intent, TreeNodeInfo, Tree } from "@blueprintjs/core";
 import { Example, ExampleProps } from "@blueprintjs/docs-theme";
-import { Tooltip2 } from "@blueprintjs/popover2";
+import { Classes as Popover2Classes, ContextMenu2, Tooltip2 } from "@blueprintjs/popover2";
 
-export interface TreeExampleState {
-    nodes: TreeNodeInfo[];
+type NodePath = number[];
+
+type TreeAction =
+    | { type: "SET_IS_EXPANDED"; payload: { path: NodePath; isExpanded: boolean } }
+    | { type: "DESELECT_ALL" }
+    | { type: "SET_IS_SELECTED"; payload: { path: NodePath; isSelected: boolean } };
+
+function forEachNode(nodes: TreeNodeInfo[], callback: (node: TreeNodeInfo) => void) {
+    if (nodes == null) {
+        return;
+    }
+
+    for (const node of nodes) {
+        callback(node);
+        forEachNode(node.childNodes, callback);
+    }
 }
 
-// use Component so it re-renders everytime: `nodes` are not a primitive type
-// and therefore aren't included in shallow prop comparison
-export class TreeExample extends React.Component<ExampleProps, TreeExampleState> {
-    public state: TreeExampleState = { nodes: INITIAL_STATE };
+function forNodeAtPath(nodes: TreeNodeInfo[], path: NodePath, callback: (node: TreeNodeInfo) => void) {
+    callback(Tree.nodeFromPath(path, nodes));
+}
 
-    public render() {
-        return (
-            <Example options={false} {...this.props}>
-                <Tree
-                    contents={this.state.nodes}
-                    onNodeClick={this.handleNodeClick}
-                    onNodeCollapse={this.handleNodeCollapse}
-                    onNodeExpand={this.handleNodeExpand}
-                    className={Classes.ELEVATION_0}
-                />
-            </Example>
-        );
+function treeExampleReducer(state: TreeNodeInfo[], action: TreeAction) {
+    switch (action.type) {
+        case "DESELECT_ALL":
+            const newState1 = [...state];
+            forEachNode(newState1, node => (node.isSelected = false));
```

Comment:
Doesn't this mutate the non-root nodes since you only did a shallow copy?

Reply:
yep, good catch, I noticed this too, I will have to do a deep copy for this to be pure / correct